File: ios/SharedList/Core/Auth/TokenStore.swift
import Foundation
import Observation

// TokenStore is the single owner of the access + refresh token pair.
//
// Two concerns rolled into one type, intentionally:
//   1. Persistence — read/write the pair to KeychainStore, survive app relaunches.
//   2. Observation  — expose the current pair to UI / APIClient as @Observable
//      state so screens can react to login/logout without an extra notification
//      bus. Logging out is "set tokens to nil"; the RootView's switch on
//      `tokens == nil` does the rest.
//
// We use `@Observable` (Observation framework, iOS 17+) instead of @ObservableObject.
// Two practical reasons for this project:
//   - It works with classes (not structs/actors), and ergonomically lets us
//     treat the store as a single shared instance threaded through AppContainer.
//   - The compiler-synthesized observation tracking is per-property, so a view
//     that reads only `currentUser` doesn't re-render when `accessToken` rotates.
//
// Threading: the type is `@MainActor`-isolated. Reasoning: the @Observable
// surface is read directly by SwiftUI views, which are MainActor-bound, so any
// mutation must happen on the main actor anyway to avoid SwiftUI's "modifying
// state during view update" warnings. The Keychain calls underneath are
// `await`-ed and run their I/O off-main implicitly via async; the cost is one
// hop per call which is fine for tokens-on-launch and tokens-on-refresh.

@MainActor
@Observable
public final class TokenStore {
    public struct Tokens: Equatable, Sendable {
        public let accessToken: String
        public let refreshToken: String
        public let user: AuthUser

        public init(accessToken: String, refreshToken: String, user: AuthUser) {
            self.accessToken = accessToken
            self.refreshToken = refreshToken
            self.user = user
        }
    }

    // Non-observed because we only read it once at init; the user could
    // subscribe to changes if we ever change the keychain at runtime, which
    // we never do.
    @ObservationIgnored
    private let keychain: any KeychainStoring

    // The single piece of observable state. nil = logged out. Setting it to a
    // value means "logged in"; APIClient picks up the new accessToken on its
    // next request (it reads `current?.accessToken` lazily, not at init).
    public private(set) var current: Tokens?

    // Keys are namespaced under the app's Keychain partition (KeychainStore's
    // service), so collisions with other apps are impossible. The strings
    // here are stable wire identifiers — never rename without a migration
    // (which we won't bother with for v1; the worst case is "log in again").
    private enum Key {
        static let accessToken = "auth.accessToken"
        static let refreshToken = "auth.refreshToken"
        static let userId = "auth.user.id"
        static let userEmail = "auth.user.email"
        static let userDisplayName = "auth.user.displayName"
    }

    public init(keychain: any KeychainStoring) {
        self.keychain = keychain
    }

    // Called once at app launch from AppContainer. We hydrate the in-memory
    // `current` pair from the Keychain so the app comes up logged-in if a
    // previous session left tokens behind. Failure to load (corrupted entry,
    // partial write from a crash) is treated as "no tokens" — safe default.
    public func loadFromKeychain() async {
        do {
            guard let access = try await keychain.get(Key.accessToken),
                  let refresh = try await keychain.get(Key.refreshToken),
                  let userId = try await keychain.get(Key.userId),
                  let email = try await keychain.get(Key.userEmail),
                  let displayName = try await keychain.get(Key.userDisplayName) else {
                current = nil
                return
            }
            current = Tokens(
                accessToken: access,
                refreshToken: refresh,
                user: AuthUser(id: userId, email: email, displayName: displayName)
            )
        } catch {
            // A keychain read error at launch is rare. Treat it as "logged
            // out" rather than crashing — a re-login fixes it. The error is
            // not silently dropped at the system level (OSStatus is logged
            // by the OS); we just don't propagate it to UI.
            current = nil
        }
    }

    // Persist a fresh token pair after signup / login / refresh. The
    // in-memory `current` is updated first so any view binding sees the new
    // identity immediately; the Keychain writes happen after but errors there
    // are still surfaced — we don't want a half-saved state to silently
    // succeed because then a relaunch would log the user back out.
    public func save(_ tokens: Tokens) async throws {
        current = tokens
        try await keychain.set(tokens.accessToken, for: Key.accessToken)
        try await keychain.set(tokens.refreshToken, for: Key.refreshToken)
        try await keychain.set(tokens.user.id, for: Key.userId)
        try await keychain.set(tokens.user.email, for: Key.userEmail)
        try await keychain.set(tokens.user.displayName, for: Key.userDisplayName)
    }

    // After a refresh response, only the access + refresh tokens change; the
    // user identity is identical. Avoid rewriting unchanged keychain rows.
    public func updateTokens(accessToken: String, refreshToken: String) async throws {
        guard let user = current?.user else {
            // Programmer error: refreshing tokens with no user record means
            // someone called this from the wrong path. Fail loudly.
            throw TokenStoreError.notLoggedIn
        }
        let updated = Tokens(accessToken: accessToken, refreshToken: refreshToken, user: user)
        current = updated
        try await keychain.set(accessToken, for: Key.accessToken)
        try await keychain.set(refreshToken, for: Key.refreshToken)
    }

    public func clear() async {
        current = nil
        // Best-effort delete of every key. We use try? because logout must
        // succeed from the user's perspective even if a keychain row was
        // somehow already missing — relaunch will recompute "logged out"
        // from the absence of any one of these.
        try? await keychain.delete(Key.accessToken)
        try? await keychain.delete(Key.refreshToken)
        try? await keychain.delete(Key.userId)
        try? await keychain.delete(Key.userEmail)
        try? await keychain.delete(Key.userDisplayName)
    }
}

public enum TokenStoreError: Error, Equatable, Sendable {
    case notLoggedIn
}

// Helpers used by APIClient to plumb a /auth/refresh response back into the
// store. They live in the same file as TokenStore so they can mutate the
// `private(set)` `current` property; APIClient.swift is in a different file
// and would only see the public read accessor.
extension TokenStore {
    // Apply a rotated pair to the in-memory state immediately. The Keychain
    // is written separately (via `persistRotated`) after the retried request
    // fires — splitting the two halves means a slow keychain write doesn't
    // delay the retry, which is the user-visible work.
    func applyRefresh(accessToken: String, refreshToken: String) throws {
        guard let user = current?.user else {
            throw TokenStoreError.notLoggedIn
        }
        current = Tokens(accessToken: accessToken, refreshToken: refreshToken, user: user)
    }

    // Persist the rotated pair. Failure is non-fatal — worst case is a
    // relaunch reads the older tokens and triggers another refresh.
    func persistRotated(accessToken: String, refreshToken: String) async throws {
        try await updateTokens(accessToken: accessToken, refreshToken: refreshToken)
    }
}

File: android/app/src/main/java/in/santosh_bharadwaj/sharedlist/core/auth/TokenStore.kt
package `in`.santosh_bharadwaj.sharedlist.core.auth

import `in`.santosh_bharadwaj.sharedlist.core.storage.SecureStorage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * Single owner of the access + refresh token pair, mirroring iOS `TokenStore`.
 *
 * Two concerns rolled into one type, intentionally:
 *   1. Persistence — read/write the pair to SecureStorage, survive process death.
 *   2. Observation — expose the current pair as `StateFlow<Tokens?>` so screens
 *      can react to login/logout without an extra event bus. Logging out =
 *      "set tokens to null"; the RootScreen's switch on `tokens == null` does
 *      the rest.
 *
 * Why `StateFlow` rather than `LiveData` or a callback?
 *   - It's the modern Kotlin-coroutines-native primitive (vs LiveData which
 *     lives in androidx.lifecycle and is observed via Java patterns).
 *   - It composes cleanly with `collectAsStateWithLifecycle()` in Compose —
 *     one line, automatic resubscription on lifecycle, no leaks.
 *   - `MutableStateFlow.update { }` (used below) gives us atomic
 *     compare-and-set semantics on the contained value, which matters when
 *     two coroutines could race to mutate the token pair (e.g. concurrent
 *     refresh attempts in the brief window before single-flight serializes
 *     them).
 *
 * Threading: `StateFlow` is itself thread-safe (it uses CAS internally). The
 * SecureStorage I/O is wrapped in a Mutex so two simultaneous writes can't
 * interleave key-by-key — partial writes are the worst kind of corruption
 * because they survive process death and look "logged in" until something
 * tries to use the broken pair.
 */
public class TokenStore(
    private val storage: SecureStorage,
) {
    public data class Tokens(
        public val accessToken: String,
        public val refreshToken: String,
        public val user: AuthUser,
    )

    private val _state = MutableStateFlow<Tokens?>(null)

    /** Observable token state. `null` = signed out; non-null = signed in. */
    public val state: StateFlow<Tokens?> = _state.asStateFlow()

    /** Synchronous read for places that just need the current value (e.g. ApiClient header injection). */
    public val current: Tokens? get() = _state.value

    /** Serializes write paths through SecureStorage — see class-level note. */
    private val ioMutex = Mutex()

    private object Key {
        const val ACCESS_TOKEN = "auth.accessToken"
        const val REFRESH_TOKEN = "auth.refreshToken"
        const val USER_ID = "auth.user.id"
        const val USER_EMAIL = "auth.user.email"
        const val USER_DISPLAY_NAME = "auth.user.displayName"
    }

    /**
     * Hydrate the in-memory state from secure storage. Called once at process
     * start from `AppContainer.bootstrap()`. A failed read (corrupted entry,
     * partial write from a crashed previous run, KeyStore reset) is treated
     * as "no tokens" — safe default; user can re-login.
     */
    public suspend fun loadFromStorage() {
        ioMutex.withLock {
            try {
                val access = storage.get(Key.ACCESS_TOKEN)
                val refresh = storage.get(Key.REFRESH_TOKEN)
                val userId = storage.get(Key.USER_ID)
                val email = storage.get(Key.USER_EMAIL)
                val displayName = storage.get(Key.USER_DISPLAY_NAME)

                if (access != null && refresh != null && userId != null && email != null && displayName != null) {
                    _state.value = Tokens(
                        accessToken = access,
                        refreshToken = refresh,
                        user = AuthUser(id = userId, email = email, displayName = displayName),
                    )
                } else {
                    _state.value = null
                }
            } catch (t: Throwable) {
                // EncryptedSharedPreferences can throw if the master key is
                // corrupted. Treat as logged-out.
                _state.value = null
            }
        }
    }

    /** Persist a fresh token pair after signup / login. In-memory state is
     *  updated FIRST so any UI binding sees the new identity immediately. */
    public suspend fun save(tokens: Tokens) {
        _state.value = tokens
        ioMutex.withLock {
            storage.set(Key.ACCESS_TOKEN, tokens.accessToken)
            storage.set(Key.REFRESH_TOKEN, tokens.refreshToken)
            storage.set(Key.USER_ID, tokens.user.id)
            storage.set(Key.USER_EMAIL, tokens.user.email)
            storage.set(Key.USER_DISPLAY_NAME, tokens.user.displayName)
        }
    }

    /**
     * After a refresh response, only access + refresh change; the user
     * identity is identical. Avoids rewriting unchanged keys. Throws if
     * called while logged out — that's a programmer error, not a runtime
     * condition.
     */
    public suspend fun updateTokens(accessToken: String, refreshToken: String) {
        val user = current?.user ?: throw TokenStoreException.NotLoggedIn
        val updated = Tokens(accessToken = accessToken, refreshToken = refreshToken, user = user)
        _state.value = updated
        ioMutex.withLock {
            storage.set(Key.ACCESS_TOKEN, accessToken)
            storage.set(Key.REFRESH_TOKEN, refreshToken)
        }
    }

    /**
     * Apply a rotated pair to the in-memory state immediately, without
     * waiting for the storage write. Used by ApiClient on the refresh path:
     * the retried request needs the new access token in memory ASAP, while
     * the storage write can happen in the background. Failure to persist is
     * non-fatal — worst case a relaunch reads the old tokens and triggers
     * another refresh, which is fine (idempotent).
     */
    public fun applyRefresh(accessToken: String, refreshToken: String) {
        _state.update { tokens ->
            val user = tokens?.user ?: throw TokenStoreException.NotLoggedIn
            Tokens(accessToken = accessToken, refreshToken = refreshToken, user = user)
        }
    }

    /** Persist the rotated pair after [applyRefresh]. Failure-tolerant. */
    public suspend fun persistRotated(accessToken: String, refreshToken: String) {
        ioMutex.withLock {
            try {
                storage.set(Key.ACCESS_TOKEN, accessToken)
                storage.set(Key.REFRESH_TOKEN, refreshToken)
            } catch (t: Throwable) {
                // Intentionally swallowed — see kdoc on applyRefresh.
            }
        }
    }

    /**
     * Clear all tokens. Best-effort delete of every key — logout must succeed
     * from the user's perspective even if a storage row was somehow missing.
     */
    public suspend fun clear() {
        _state.value = null
        ioMutex.withLock {
            runCatching { storage.delete(Key.ACCESS_TOKEN) }
            runCatching { storage.delete(Key.REFRESH_TOKEN) }
            runCatching { storage.delete(Key.USER_ID) }
            runCatching { storage.delete(Key.USER_EMAIL) }
            runCatching { storage.delete(Key.USER_DISPLAY_NAME) }
        }
    }
}

public sealed class TokenStoreException(message: String) : Exception(message) {
    public object NotLoggedIn : TokenStoreException("token rotation requested while logged out") {
        // readResolve is called reflectively by Java serialization to ensure
        // the singleton stays a singleton across deserialization. Detekt
        // flags this as unused because it's never called from Kotlin code.
        @Suppress("unused")
        private fun readResolve(): Any = NotLoggedIn
    }
}
